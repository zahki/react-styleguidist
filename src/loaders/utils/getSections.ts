// This two functions should be in the same file because of cyclic imports

import fs from 'fs';
import path from 'path';
import _ from 'lodash';
import requireIt from './requireIt';
import getComponentFiles from './getComponentFiles';
import getComponents from './getComponents';
import slugger from './slugger';
import * as Rsg from '../../typings';

const examplesLoader = path.resolve(__dirname, '../examples-loader.js');

function processSectionContent(
	section: Rsg.ConfigSection,
	config: Rsg.SanitizedStyleguidistConfig
): Rsg.RequireItResult | Rsg.MarkdownExample | undefined {
	if (!section.content) {
		return undefined;
	}

	const contentRelativePath = section.content;

	if (_.isFunction(section.content)) {
		return {
			type: 'markdown',
			content: section.content(),
		};
	}

	// Try to load section content file
	const contentAbsolutePath = path.resolve(config.configDir, contentRelativePath);
	if (!fs.existsSync(contentAbsolutePath)) {
		throw new Error(`Styleguidist: Section content file not found: ${contentAbsolutePath}`);
	}
	return requireIt(`!!${examplesLoader}!${contentAbsolutePath}`);
}

const getSectionComponents = (
	section: Rsg.ConfigSection,
	config: Rsg.SanitizedStyleguidistConfig
) => {
	let ignore = config.ignore ? _.castArray(config.ignore) : [];
	if (section.ignore) {
		ignore = ignore.concat(_.castArray(section.ignore));
	}

	return getComponents(getComponentFiles(section.components, config.configDir, ignore), config);
};

/**
 * Return object for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @param {number} parentDepth
 * @returns {Array}
 */
export default function getSections(
	sections: Rsg.ConfigSection[],
	config: Rsg.SanitizedStyleguidistConfig,
	parentDepth?: number
): Rsg.LoaderSection[] {
	// eslint-disable-next-line @typescript-eslint/no-use-before-define
	return sections.map(section => processSection(section, config, parentDepth));
}

/**
 * Return an object for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @param {number} parentDepth
 * @returns {object}
 */
export function processSection(
	section: Rsg.ConfigSection,
	config: Rsg.SanitizedStyleguidistConfig,
	parentDepth?: number
): Rsg.LoaderSection {
	const content = processSectionContent(section, config);

	let sectionDepth;

	if (parentDepth === undefined) {
		sectionDepth = section.sectionDepth !== undefined ? section.sectionDepth : 0;
	} else {
		sectionDepth = parentDepth === 0 ? 0 : parentDepth - 1;
	}

	return {
		...section,
		exampleMode: section.exampleMode || config.exampleMode,
		usageMode: section.usageMode || config.usageMode,
		sectionDepth,
		slug: `section-${slugger.slug(section.name || 'untitled')}`,
		sections: getSections(section.sections || [], config, sectionDepth),
		href: section.href,
		components: getSectionComponents(section, config),
		content,
	};
}
