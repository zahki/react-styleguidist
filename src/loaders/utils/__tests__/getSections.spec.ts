import path from 'path';
import getSections, { processSection } from '../getSections';
import * as Rsg from '../../../typings';

const configDir = path.resolve(__dirname, '../../../../test');
const config = {
	configDir,
	exampleMode: 'collapse',
	usageMode: 'collapse',
	getExampleFilename: (a: string) => a,
	getComponentPathLine: (a: string) => a,
} as Rsg.SanitizedStyleguidistConfig;

const sections: Rsg.ConfigSection[] = [
	{
		name: 'Readme',
		content: 'components/Button/Readme.md',
	},
	{
		name: 'Components',
		components: 'components/**/[A-Z]*.js',
	},
	{
		name: 'Ignore',
		components: 'components/**/*.js',
		ignore: '**/components/Annotation/*',
	},
	{
		name: 'Ignore',
		content: () => 'Hello World',
	} as any,
];
const sectionsWithDepth = [
	{
		name: 'Documentation',
		sections: [
			{
				name: 'Files',
				sections: [
					{
						name: 'First File',
					},
				],
			},
		],
		sectionDepth: 2,
	},
	{
		name: 'Components',
		expand: true,
		sections: [
			{
				name: 'Buttons',
			},
		],
		sectionDepth: 0,
	},
];
const sectionsWithBadDepth = [
	{
		name: 'Documentation',
		sections: [
			{
				name: 'Files',
				sections: [
					{
						name: 'First File',
					},
				],
				sectionDepth: 2,
			},
		],
	},
];

function filterSectionDepth(section: Rsg.LoaderSection): Rsg.ConfigSection {
	if (section.sections && section.sections.length) {
		return {
			sectionDepth: section.sectionDepth,
			sections: section.sections.map(filterSectionDepth),
		};
	}
	return {
		sectionDepth: section.sectionDepth,
	};
}

it('processSection() should return an object for section with content', () => {
	const result = processSection(sections[0], config);

	expect(result).toMatchSnapshot();
});

it('processSection() should throw when content file not found', () => {
	const fn = () => processSection({ content: 'pizza' }, config);

	expect(fn).toThrowError('Section content file not found');
});

it('processSection() should return an object for section with components', () => {
	const result = processSection(sections[1], config);

	expect(result).toMatchSnapshot();
});

it('processSection() should return an object for section without ignored components', () => {
	const result = processSection(sections[2], config);

	expect(result).toMatchSnapshot();
});

it('processSection() should return an object for section with content as function', () => {
	const result = processSection(sections[3], config);

	expect(result).toMatchSnapshot();
});

it('getSections() should return an array', () => {
	const result = getSections(sections, config);

	expect(result).toMatchSnapshot();
});

it('getSections() should return an array of sectionsWithDepth with sectionDepth decreasing', () => {
	const result = getSections(sectionsWithDepth, config);

	expect(result.map(filterSectionDepth)).toEqual([
		{
			sectionDepth: 2,
			sections: [
				{
					sectionDepth: 1,
					sections: [
						{
							sectionDepth: 0,
						},
					],
				},
			],
		},
		{
			sectionDepth: 0,
			sections: [
				{
					sectionDepth: 0,
				},
			],
		},
	]);
});

it('getSections() should make custom options by user available', () => {
	const result = getSections(sectionsWithDepth, config);
	const expandSection = result.find(section => section.name === 'Components');
	expect(expandSection).toHaveProperty('expand');
});

it('getSections() should return an array of sectionsWithBadDepth taking the sectionDepth of the first depth of the sections', () => {
	const result = getSections(sectionsWithBadDepth, config);

	expect(result.map(filterSectionDepth)).toEqual([
		{
			sectionDepth: 0,
			sections: [
				{
					sectionDepth: 0,
					sections: [
						{
							sectionDepth: 0,
						},
					],
				},
			],
		},
	]);
});
