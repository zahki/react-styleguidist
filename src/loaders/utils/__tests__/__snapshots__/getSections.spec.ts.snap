// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSections() should return an array 1`] = `
Array [
  Object {
    "components": Array [],
    "content": Object {
      "require": "!!~/src/loaders/examples-loader.js!~/test/components/Button/Readme.md",
    },
    "exampleMode": "collapse",
    "href": undefined,
    "name": "Readme",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "section-readme-1",
    "usageMode": "collapse",
  },
  Object {
    "components": Array [
      Object {
        "filepath": "components/Annotation/Annotation.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Annotation/Annotation.js",
        },
        "pathLine": "components/Annotation/Annotation.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Annotation/Annotation.js",
        },
        "slug": "annotation-1",
      },
      Object {
        "filepath": "components/Button/Button.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Button/Button.js",
        },
        "pathLine": "components/Button/Button.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Button/Button.js",
        },
        "slug": "button-2",
      },
      Object {
        "filepath": "components/Placeholder/Placeholder.js",
        "hasExamples": true,
        "metadata": Object {
          "require": "~/test/components/Placeholder/Placeholder.json",
        },
        "module": Object {
          "require": "~/test/components/Placeholder/Placeholder.js",
        },
        "pathLine": "components/Placeholder/Placeholder.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Placeholder/Placeholder.js",
        },
        "slug": "placeholder-2",
      },
      Object {
        "filepath": "components/Price/Price.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Price/Price.js",
        },
        "pathLine": "components/Price/Price.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Price/Price.js",
        },
        "slug": "price-2",
      },
      Object {
        "filepath": "components/RandomButton/RandomButton.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/RandomButton/RandomButton.js",
        },
        "pathLine": "components/RandomButton/RandomButton.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/RandomButton/RandomButton.js",
        },
        "slug": "randombutton-2",
      },
    ],
    "content": undefined,
    "exampleMode": "collapse",
    "href": undefined,
    "name": "Components",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "section-components-1",
    "usageMode": "collapse",
  },
  Object {
    "components": Array [
      Object {
        "filepath": "components/Button/Button.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Button/Button.js",
        },
        "pathLine": "components/Button/Button.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Button/Button.js",
        },
        "slug": "button-3",
      },
      Object {
        "filepath": "components/Label/index.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Label/index.js",
        },
        "pathLine": "components/Label/index.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Label/index.js",
        },
        "slug": "label-1",
      },
      Object {
        "filepath": "components/Placeholder/Placeholder.js",
        "hasExamples": true,
        "metadata": Object {
          "require": "~/test/components/Placeholder/Placeholder.json",
        },
        "module": Object {
          "require": "~/test/components/Placeholder/Placeholder.js",
        },
        "pathLine": "components/Placeholder/Placeholder.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Placeholder/Placeholder.js",
        },
        "slug": "placeholder-3",
      },
      Object {
        "filepath": "components/Price/Price.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Price/Price.js",
        },
        "pathLine": "components/Price/Price.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/Price/Price.js",
        },
        "slug": "price-3",
      },
      Object {
        "filepath": "components/RandomButton/RandomButton.js",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/RandomButton/RandomButton.js",
        },
        "pathLine": "components/RandomButton/RandomButton.js",
        "props": Object {
          "require": "!!~/src/loaders/props-loader.js!~/test/components/RandomButton/RandomButton.js",
        },
        "slug": "randombutton-3",
      },
    ],
    "content": undefined,
    "exampleMode": "collapse",
    "href": undefined,
    "ignore": "**/components/Annotation/*",
    "name": "Ignore",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "section-ignore-2",
    "usageMode": "collapse",
  },
  Object {
    "components": Array [],
    "content": Object {
      "content": "Hello World",
      "type": "markdown",
    },
    "exampleMode": "collapse",
    "href": undefined,
    "name": "Ignore",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "section-ignore-3",
    "usageMode": "collapse",
  },
]
`;

exports[`processSection() should return an object for section with components 1`] = `
Object {
  "components": Array [
    Object {
      "filepath": "components/Annotation/Annotation.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Annotation/Annotation.js",
      },
      "pathLine": "components/Annotation/Annotation.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Annotation/Annotation.js",
      },
      "slug": "annotation",
    },
    Object {
      "filepath": "components/Button/Button.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Button/Button.js",
      },
      "pathLine": "components/Button/Button.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Button/Button.js",
      },
      "slug": "button",
    },
    Object {
      "filepath": "components/Placeholder/Placeholder.js",
      "hasExamples": true,
      "metadata": Object {
        "require": "~/test/components/Placeholder/Placeholder.json",
      },
      "module": Object {
        "require": "~/test/components/Placeholder/Placeholder.js",
      },
      "pathLine": "components/Placeholder/Placeholder.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Placeholder/Placeholder.js",
      },
      "slug": "placeholder",
    },
    Object {
      "filepath": "components/Price/Price.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Price/Price.js",
      },
      "pathLine": "components/Price/Price.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Price/Price.js",
      },
      "slug": "price",
    },
    Object {
      "filepath": "components/RandomButton/RandomButton.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/RandomButton/RandomButton.js",
      },
      "pathLine": "components/RandomButton/RandomButton.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/RandomButton/RandomButton.js",
      },
      "slug": "randombutton",
    },
  ],
  "content": undefined,
  "exampleMode": "collapse",
  "href": undefined,
  "name": "Components",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "section-components",
  "usageMode": "collapse",
}
`;

exports[`processSection() should return an object for section with content 1`] = `
Object {
  "components": Array [],
  "content": Object {
    "require": "!!~/src/loaders/examples-loader.js!~/test/components/Button/Readme.md",
  },
  "exampleMode": "collapse",
  "href": undefined,
  "name": "Readme",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "section-readme",
  "usageMode": "collapse",
}
`;

exports[`processSection() should return an object for section with content as function 1`] = `
Object {
  "components": Array [],
  "content": Object {
    "content": "Hello World",
    "type": "markdown",
  },
  "exampleMode": "collapse",
  "href": undefined,
  "name": "Ignore",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "section-ignore-1",
  "usageMode": "collapse",
}
`;

exports[`processSection() should return an object for section without ignored components 1`] = `
Object {
  "components": Array [
    Object {
      "filepath": "components/Button/Button.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Button/Button.js",
      },
      "pathLine": "components/Button/Button.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Button/Button.js",
      },
      "slug": "button-1",
    },
    Object {
      "filepath": "components/Label/index.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Label/index.js",
      },
      "pathLine": "components/Label/index.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Label/index.js",
      },
      "slug": "label",
    },
    Object {
      "filepath": "components/Placeholder/Placeholder.js",
      "hasExamples": true,
      "metadata": Object {
        "require": "~/test/components/Placeholder/Placeholder.json",
      },
      "module": Object {
        "require": "~/test/components/Placeholder/Placeholder.js",
      },
      "pathLine": "components/Placeholder/Placeholder.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Placeholder/Placeholder.js",
      },
      "slug": "placeholder-1",
    },
    Object {
      "filepath": "components/Price/Price.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Price/Price.js",
      },
      "pathLine": "components/Price/Price.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/Price/Price.js",
      },
      "slug": "price-1",
    },
    Object {
      "filepath": "components/RandomButton/RandomButton.js",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/RandomButton/RandomButton.js",
      },
      "pathLine": "components/RandomButton/RandomButton.js",
      "props": Object {
        "require": "!!~/src/loaders/props-loader.js!~/test/components/RandomButton/RandomButton.js",
      },
      "slug": "randombutton-1",
    },
  ],
  "content": undefined,
  "exampleMode": "collapse",
  "href": undefined,
  "ignore": "**/components/Annotation/*",
  "name": "Ignore",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "section-ignore",
  "usageMode": "collapse",
}
`;
